src/cmds/qdel.c

```diff
@@ -58,106 +58,212 @@
 #include <pbs_config.h>   /* the master config generated by configure */
 #include <pbs_version.h>
 
-#define NO_OF_JOBIDS 1000 /* Using this macro to allocate array of pointers */
+#define MAX_TIME_DELAY_LEN 32
+#define GETOPT_ARGS "W:x"
 
-typedef struct svr_jobid_list svr_jobid_list_t;
-struct svr_jobid_list{
-	int max_sz;
-	int total_jobs;
-	char *server_name;
-	char **jobids;
-	svr_jobid_list_t *next;
-};
+extern void free_svrjobidlist(svr_jobid_list_t *list);
+extern int append_jobid(svr_jobid_list_t *svr, char *jobid);
+extern int add_jid_to_list_by_name(char *job_id, char *svrname, svr_jobid_list_t **svr_jobid_list_hd);
 
-/**
- * @brief
- *	Append a given jobid to the given svr_jobid_list struct.
- *
- * @param[in] svr - Server name
- * @param[in] jobid - Job id 
- * 
- * @return void
- */
-void
-append_jobid(svr_jobid_list_t *svr, char *jobid)
+static int
+process_deljobstat(char *clusterid, struct batch_deljob_status *list, svr_jobid_list_t **rmtlist)
+{
+	struct batch_deljob_status *p_delstatus;
+	char *errtxt = NULL;
+	int any_failed = 0;
+
+	for (p_delstatus = list; p_delstatus != NULL; p_delstatus = p_delstatus->next) {
+		if (p_delstatus->code == PBSE_UNKJOBID && rmtlist != NULL) {
+			char rmt_server[PBS_MAXDEST + 1];
+
+			/* Check if job was moved to a remote cluster */
+			if (locate_job(p_delstatus->name, clusterid, rmt_server))
+				add_jid_to_list_by_name(p_delstatus->name, rmt_server, rmtlist);
+
+		} else if (p_delstatus->code != PBSE_HISTJOBDELETED) {
+			errtxt = pbse_to_txt(p_delstatus->code);
+			if ((errtxt != NULL) && (p_delstatus->code != PBSE_HISTJOBDELETED)) {
+				fprintf(stderr, "%s: %s %s\n", "qdel", errtxt, p_delstatus->name);
+				any_failed = p_delstatus->code;
+			}
+		}
+	}
+
+	return any_failed;
+}
+
+static int
+get_mail_suppress_count(int connect, int dfltmail, int dfltmailflg)
 {
-	
-	if ((svr == NULL) || (jobid == NULL))
-		return;
-	
-	if (svr->max_sz == 0) {
-		svr->jobids = malloc((NO_OF_JOBIDS + 1) * sizeof(char *));
-		if (svr->jobids == NULL) 
-			goto error;
-		svr->max_sz = NO_OF_JOBIDS;
-	} else if (svr->total_jobs == svr->max_sz) {
-		svr->max_sz *= 2;
-		svr->jobids = realloc(svr->jobids, (svr->max_sz + 1) * sizeof(char *));
-		if (svr->jobids == NULL) 
-			goto error;
+	struct batch_status *ss = NULL;
+	struct batch_status *iter_ss = NULL;
+	char *errmsg;
+	char *keystr, *valuestr;
+
+	if (dfltmailflg == FALSE) {
+		ss = pbs_statserver(connect, NULL, NULL);
+		if (ss == NULL && pbs_errno != PBSE_NONE) {
+			if ((errmsg = pbs_geterrmsg(connect)) != NULL)
+				fprintf(stderr, "qdel: %s\n", errmsg);
+			else
+				fprintf(stderr, "qdel: Error %d\n", pbs_errno);
+			exit(pbs_errno);
+		}
+
+		for (iter_ss = ss; iter_ss != NULL; iter_ss = iter_ss->next) {
```

Comment:
> I assume this loop is to loop on multiple servers.
If that is the case then shouldn't we be setting dfltmail for each corresponding server? Can't two different servers have different values of default qdel args?

No I think this loop was a mistake, we will never have more than one object returned by statserver. Also, we use 'ss' instead of 'iter_ss' inside the loop so it was only looking at the first member of the list.

I call this function from "delete_jobs_for_cluster", which is called for every cluster (or every server, pre multi-server), so it should consider each cluster's unique dflt_qdel_args.